relay-server/src/services/project/state/fetch_state.rs

```diff
@@ -0,0 +1,145 @@
+use std::sync::Arc;
+
+use relay_config::Config;
+use relay_dynamic_config::ProjectConfig;
+use tokio::time::Instant;
+
+use crate::services::project::state::info::ProjectInfo;
+use crate::services::project::ProjectState;
+
+/// Hides a cached project state and only exposes it if it has not expired.
+#[derive(Clone, Debug)]
+pub struct ProjectFetchState {
+    /// The time at which this project state was last updated.
+    last_fetch: Option<Instant>,
+    state: ProjectState,
+}
+
+impl ProjectFetchState {
+    /// Takes a [`ProjectState`] and sets it's last fetch to the current time.
+    pub fn new(state: ProjectState) -> Self {
+        Self {
+            last_fetch: Some(Instant::now()),
+            state,
+        }
+    }
+
+    /// Project state for an unknown but allowed project.
+    ///
+    /// This state is used for forwarding in Proxy mode.
+    pub fn allowed() -> Self {
+        Self::enabled(ProjectInfo {
+            project_id: None,
+            last_change: None,
+            public_keys: Default::default(),
+            slug: None,
+            config: ProjectConfig::default(),
+            organization_id: None,
+        })
+    }
+
+    /// An enabled project state created from a project info.
+    pub fn enabled(project_info: ProjectInfo) -> Self {
+        Self::new(ProjectState::Enabled(Arc::new(project_info)))
+    }
+
+    // Returns a disabled state.
+    pub fn disabled() -> Self {
+        Self::new(ProjectState::Disabled)
+    }
+
+    /// Returns a pending or invalid state.
+    pub fn pending() -> Self {
+        Self::new(ProjectState::Pending)
+    }
+
+    /// Create a config that immediately counts as expired.
+    ///
+    /// This is what [`super::Project`] initializes itself with.
+    pub fn expired() -> Self {
+        Self {
+            // Make sure the state immediately qualifies as expired:
+            last_fetch: None,
+            state: ProjectState::Pending,
+        }
+    }
+
+    /// Sanitizes the contained project state. See [`ProjectState::sanitize`].
+    pub fn sanitize(self) -> Self {
```

Comment:
```suggestion
    pub fn sanitized(self) -> Self {
```

Seems more like a transition than a mutation